src/external/wannier90.jl

```diff
@@ -287,3 +289,79 @@ default_wannier_centres(n_wannier) = [rand(1, 3) for _ in 1:n_wannier]
     @timing "Wannierization" wannier90_jll.wannier90(exe -> run(Cmd(`$exe $prefix`; dir)))
     fileprefix
 end
+
+@doc raw"""
+    For a given set of bands ``\{Ψ_{nk}\}_{1\leq n \leq N_in}``
+    and a Wannier90 file containing unitary matrices (for disentanglement or
+    Wannierization) returns a set of band ``{Ψ_out_{nk}}_{1\leq n \leq N_out}``
+    with:
+    ``Ψ_out_{nk} = \sum\limits_{m=1}^{N_in} U_{mn}Ψ_{nk}``
+    ``N_in > N_out`` for disentanglement.
+    ``N_in = N_out`` for wannierization.
+"""
+function apply_U_matrices(file, ψ)
+    # Extract all Uk
+    @assert(isfile(file))
+    Uks = readdlm(file); num_kpts, num_ψout, num_ψin = Uks[2,1:3]
+    len_Uk = num_ψin*num_ψout
+
+    ψ_out = copy(ψ) # Better ideas than copy to inherit the type of ψ in ψ_out ?
+    for k in 1:num_kpts
+        # Locate the beginning of the matrix Uk written as a vector
+        i_Uk = 4 + (k-1)*(len_Uk+1)
+        Uk = reshape([x+im*y for (x,y) in eachrow(Uks[i_Uk:i_Uk+(len_Uk-1), 1:2])],
+                     num_ψin, num_ψout)
+        # Check that Uk is unitary
+        (norm(Uk'Uk - diagm(ones(num_ψout))) > 1e-8) && (@error("U matrix is not unitary"))
+        # Compute all ψ_out[k][:,n]
```

Comment:
would be cleaner to split into two functions: read_amn and apply_U or something. Also is this not just `psi_out = [psi[ik] * U[ik] for ik=1:Nk]`? If so you maybe don't even need a function for that

Reply:
Its a bit tricky because if the bands given to wannier90 were not the full set of bands in psi_in, the U matrices do not have the proper dimension to do psi[ik]*U[ik]
However it can be done and provide the same result by completing U by zeros in the first dimension I think !